### PR TITLE
add first 2 fn and tests

### PR DESCRIPTION
**Pull request by jianructose**
Hi team, I have added tests and first 2 fns. It passed tests on my side. lmk if not on yours.

### CODE REVIEW

**Comment by jianructose** [comment]
#10

File: pygtracker/pygtracker.py [hunk 1]
```diff
@@ -40,7 +40,31 @@ def register_courses(self, df):
...
         """
-        return None
+        course_list=[511, 512, 513, 521, 522, 523, 524, 525, 531, 532, 541, 542, 551, 552, 553, 554, 561, 562, 563, 571, 572, 573, 574, 575, 591]
+        input_courses=df.course_id.tolist()
+        if not set(input_courses).issubset(course_list): 
+            error_input = [x for x in input_courses if x not in course_list]
+            raise ValueError("Oops, your dataframe has non-MDS Course(s) DSCI " + str(error_input)[1:-1])
+            
+        
+        assessment_list = ['lab1', 'lab2', 'lab3', 'lab4', 'milestone1', 'milestone2', 'milestone3', 'milestone4', 'feedback', 'quiz1', 'quiz2']
```
Comment (tdkhanhvu) [inline pygtracker/pygtracker.py]: I think limiting the assessment to this list alone is quite rigid as you may have other assessment components as well. But given the scope of this project, you may want to leave it like this.

File: pygtracker/pygtracker.py [hunk 2]
```diff
@@ -40,7 +40,31 @@ def register_courses(self, df):
...
+        input_assess=df.iloc[:, 1].tolist()
+        if not set(input_assess).issubset(assessment_list):
+            error_input = [x for x in input_assess if x not in assessment_list]
+            raise ValueError(f"Oops, your dataframe has non-MDS assessment(s) {str(error_input)[1:-1]}")
+        
+        for _, w in df.weight.items():
```
Comment (tdkhanhvu) [inline pygtracker/pygtracker.py]: **Suggestions only**
You can optimize this in the vectorization manner, sth similar to this:

(df.weight < 0 or df.weight > 1).sum()

if the result returns > 1 it means one of the component violates the rule. I'm not sure about the syntax though.

File: pygtracker/pygtracker.py [hunk 3]
```diff
@@ -40,7 +40,31 @@ def register_courses(self, df):
...
+            if w<0 or w>1:
+                raise ValueError(f'One of the weight={w}, which is not within the range 0 to 1.')
+            
+        for course_id, sum_weight in df.groupby('course_id').sum('weight').weight.items():
```
Comment (tdkhanhvu) [inline pygtracker/pygtracker.py]: similar as above for the vectorization

File: pygtracker/pygtracker.py [hunk 4]
```diff
@@ -59,7 +83,30 @@ def record_grades(self, df):
         -------
         None
         """
-        return None
+        course_list=[511, 512, 513, 521, 522, 523, 524, 525, 531, 532, 541, 542, 551, 552, 553, 554, 561, 562, 563, 571, 572, 573, 574, 575, 591]
+        input_courses=df.course_id.tolist()
+        if not set(input_courses).issubset(course_list):
```
Comment (tdkhanhvu) [inline pygtracker/pygtracker.py]: If these are repeated, it is better to extract them to helper function. sth like

**check_if_course_not_exist**

File: pygtracker/pygtracker.py [hunk 5]
```diff
@@ -59,7 +83,30 @@ def record_grades(self, df):
...
-        return None
+        course_list=[511, 512, 513, 521, 522, 523, 524, 525, 531, 532, 541, 542, 551, 552, 553, 554, 561, 562, 563, 571, 572, 573, 574, 575, 591]
+        input_courses=df.course_id.tolist()
+        if not set(input_courses).issubset(course_list): 
+            error_input = [x for x in input_courses if x not in course_list]
+            raise ValueError("Oops, your dataframe has non-MDS Course(s) DSCI " + str(error_input)[1:-1])
+            
+          
+        
+        assessment_list = ['lab1', 'lab2', 'lab3', 'lab4', 'milestone1', 'milestone2', 'milestone3', 'milestone4', 'feedback', 'quiz1', 'quiz2']
```
Comment (tdkhanhvu) [inline pygtracker/pygtracker.py]: If these are repeated, it is better to extract them to helper function. sth like

**check_if_assessment_not_exist**

File: tests/test_pygtracker.py [hunk 6]
```diff
@@ -1,15 +1,93 @@
...
 
 def test_version():
     assert __version__ == '0.1.0'
     
 # Start tests for register_courses
 
+def test_register_courses_course_id_valid():
+    tracker=pygtracker.GradeTracker()
+    df=pd.DataFrame()
+    df['course_id']=[555]
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]: It is better to declare a separate helper function to generate the df frame that you are using here.

`def generate_df_register_courses(course_id, assessment_id, weight)`

File: tests/test_pygtracker.py [hunk 7]
```diff
@@ -1,15 +1,93 @@
...
+import pandas as pd
+import numpy as np
+from pandas._testing import assert_frame_equal
 
 def test_version():
     assert __version__ == '0.1.0'
     
 # Start tests for register_courses
 
+def test_register_courses_course_id_valid():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_register_courses_course_id_invalid():
```

File: tests/test_pygtracker.py [hunk 8]
```diff
@@ -1,15 +1,93 @@
...
+    tracker=pygtracker.GradeTracker()
+    df=pd.DataFrame()
+    df['course_id']=[555]
+    df['assessment_id']=['lab4']
+    df['weight']=1
+
+    with raises(ValueError):
+        tracker.register_courses(df)
+
+def test_register_courses_assess_valid():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_register_courses_assess_invalid():
```

File: tests/test_pygtracker.py [hunk 9]
```diff
@@ -1,15 +1,93 @@
...
+    tracker=pygtracker.GradeTracker()
+    df=pd.DataFrame()
+    df['course_id']=[522]
+    df['assessment_id']=['lab5']
+    df['weight']=1
+    with raises(ValueError):
+        tracker.register_courses(df)
+
+def test_register_courses_weight_valid_1():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_register_courses_weight_invalid_not_sum_to_one():
```

File: tests/test_pygtracker.py [hunk 10]
```diff
@@ -1,15 +1,93 @@
...
+    tracker=pygtracker.GradeTracker()
+    df=pd.DataFrame()
+    df['course_id']=[511, 511, 511, 511,]
+    df['assessment_id']=['lab4', 'lab2', 'lab1', 'lab3']
+    df['weight']=[.2, .3, .4, .2, ]
+    with raises(ValueError):
+        tracker.register_courses(df)
+
+def test_register_courses_weight_valid_2():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_register_courses_weight_invalid_negative():
```

File: tests/test_pygtracker.py [hunk 11]
```diff
@@ -1,15 +1,93 @@
...
+    df['assessment_id']=['lab4', 'lab2', 'lab1', 'lab3']
+    df['weight']=[ -.5, .1, .9, .5]
+    with raises(ValueError):
+        tracker.register_courses(df)
+
 # End tests for register_courses
 
 # Start tests for record_grades
 
+def test_record_grades_course_id_valid():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_record_grades_course_id_invalid():
```

File: tests/test_pygtracker.py [hunk 12]
```diff
@@ -1,15 +1,93 @@
...
+    df=pd.DataFrame()
+    df['course_id']=[517]
+    df['assessment_id']=['lab4']
+    df['grade']=100
+    df['student_id']=['elina']
+
+    with raises(ValueError):
+        tracker.record_grades(df)
+
+def test_record_grade_assess_valid():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_record_grade_assess_invalid():
```

File: tests/test_pygtracker.py [hunk 13]
```diff
@@ -1,15 +1,93 @@
...
+    df=pd.DataFrame()
+    df['course_id']=[571]
+    df['assessment_id']=['lab6']
+    df['grade']=100
+    df['student_id']=['elina']
+
+    with raises(ValueError):
+        tracker.record_grades(df)
+
+def test_record_grade_grade_valid():
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]:
```suggestion
def test_record_grade_grade_invalid_negative():
```

**Review by tdkhanhvu — CHANGES_REQUESTED** [review]
- Merge the code from the main
- Rename the test function to better reflect what we are testing
- Write a helper function to help you generate the dummy data
- Should include test cases to show that the courses and grades are recorded correctly. Currently I only see the failure cases

File: tests/test_pygtracker.py [hunk 14]
```diff
@@ -10,10 +10,84 @@ def test_version():
...
+
+    with raises(ValueError):
+        tracker.record_grades(df)
+
+def test_record_grade_grade_invalid_negative():
+    tracker=pygtracker.GradeTracker()
+    df=pd.DataFrame()
+    df['course_id']=[511, 511, 511, 511,]
+    df['assessment_id']=['lab4', 'lab2', 'lab1', 'lab3']
+    df['grade']=[100.1, -77, 99, 88.4]
```
Comment (tdkhanhvu) [inline tests/test_pygtracker.py]: It is better to split this into 2 separate tests:
- above 100
- negative

File: pygtracker/pygtracker.py (hunk 2 above)
Comment (jianructose) [reply pygtracker/pygtracker.py]: yep agree. i was on vectorization but realize might be some limitations on this

**Review by tdkhanhvu — APPROVED** [review]

**Pull request merged**